### ISSUE TITLE
Importing users via LDAP

### ISSUE THREAD

**Issue by DavidR-ZA** [snipe/snipe-it#1869]
#### Expected Behavior (or desired behavior if a feature request)

I set the LDAP settings in Admin>Settings>Edit Settings >LDAP Heading
I have four sites - each with own AD - so I did this four times with different LDAP server settings. 
With each setting. I go to People > LDAP. Set location and Synchronize.
I see the list of users. Each one shows as created with username, First Namer, LAstName and Email address.
I look at the data in the users table, 1301 rows displayed, All fields correctly created for each user. Looking similar to my own admin account. Permissions field is {"user":1} and activated is 1. Notes is Imported from LDAP. Company (NULL)

-----

#### Actual Behavior
In People > Current Users, the only user displayed is my account. None of the LDP imported users display.

If I create a new user manually and I go to the Manager select drop down, all the users appear.
This user shows in the People List - with a created_at value, while the LDAP imports have 0000-00-00 00:00:00 in this field.

In status on Create Asset, when I select, Ready to deploy, the list of users does not display. Green circle going around and then a Server Error 500, when I try to create the Asset.

This is different to the behaviour on the demo site.

-----

#### Please confirm you have done the following before posting your bug report:

- [x] I have enabled debug mode
- [x] I have read [checked the Common Issues page](http://docs.snipeitapp.com/common-issues.html)

-----
#### Please provide answers to these questions before posting your bug report:

- Version of Snipe-IT you're running
Snipe-IT version	v2.1.0-55-g701c16f
- What OS and web server you're running Snipe-IT on
Operating system	Ubuntu Linux 14.04.4
Apache Server version : Apache/2.4.7 (Ubunutu)
Server Built: Jan 14 2016 17:45:23

- What method you used to install Snipe-IT (install.sh, manual installation, docker, etc)
used install.sh
- If you're getting an error in your browser, include that error
No error message
- What specific Snipe-IT page you're on, and what specific element you're interacting with to trigger the error
/admin/users

- If a stacktrace is provided in the error, include that too.
- Any errors that appear in your browser's error console.
- Confirm whether the error is [reproduceable on the demo](https://snipeitapp.com/demo).
The Demo site does not allow for LDAP settings

- Include any additional information you can find in `app/storage/logs` and your webserver's logs.
- Include what you've done so far in the installation, and if you got any error messages along the way.
No error messages during installation
- Indicate whether or not you've manually edited any data directly in the database
No data has been entered into the database manually.

**Comment by snipe** [comment]
What are you seeing on the LDAP import screen when you initiate an LDAP import?

**Comment by DavidR-ZA** [comment]
It shows that users are being imported and created. If I run the same LDAP import again for a second time, it says they already exist. The import process proceeds normally.

**Comment by DavidR-ZA** [comment]
When I try to create an Asset and I select "Ready to deploy" it should show me a list of users to assign the asset to, but I just get the spinning green dots. It seems unable to read the users' table.

![image](https://cloud.githubusercontent.com/assets/17966170/13943147/dc713de0-f002-11e5-8ed2-8b89d1018d53.png)

**Comment by snipe** [comment]
But when you go to People, do you see those LDAP people in your list? If you click on Deleted Users, do you see them? If they are in the database, they should show up in either current users or deleted users, no two ways about it. 

Check your browser's debug messages on the People listing, and on the Create Asset page (per the initial ticket recommendation). I'd bet you'll see that an ajax call is giving a 500 error. We need to know what that 500 error is.

**Comment by DavidR-ZA** [comment]
When I go to people, I only see my original user and one I created manually:

![image](https://cloud.githubusercontent.com/assets/17966170/13944907/ef75c78a-f011-11e5-8f1d-0cc86e78813f.png)
No users show when I select "Show deleted users"
This is the user table - there are 1303 rows :
![image](https://cloud.githubusercontent.com/assets/17966170/13945028/d981ca54-f012-11e5-84e1-68a28f54e391.png)

When I look at the elements tab in the devtools panel on chrome, I do not see any errors. I can point to elements on the page like the names and I see the calls to the database but no errors.
How do I track the 500 error?
DavidR

**Comment by snipe** [comment]
You can look at your app error log, per the Getting Help documentation. The elements tab won't tell you anything in Chrome, you want the Console tab, (also per the Getting Help documentation).

**Comment by DavidR-ZA** [comment]
Thank you for taking time to help me with this. I really appreciate what you are doing here.
This is the only error I could find. It was on the "add an asset" page and it was when I selected "ready to deploy" on the Status drop down.
![image](https://cloud.githubusercontent.com/assets/17966170/13945681/e97d65e4-f017-11e5-949a-2c7230463dd9.png)

No errors show on the console when I go to the "people" page.
I have attached the log-apache2handler-2016-03-20.txt file.
[snipelog.txt](https://github.com/snipe/snipe-it/files/183731/snipelog.txt)
This was before I set debug=true in the app.php file.
Nothing new is going into today's log since I set the debug=true.

These are the queries running on the /admin/users page:
select * from `users` where `users`.`deleted_at` is null and `users`.`id` = '1' limit 1580μs
select * from `throttle` where `user_id` = '1' limit 1300μs
select `groups`.*, `users_groups`.`user_id` as `pivot_user_id`, `users_groups`.`group_id` as `pivot_group_id` from `groups` inner join `users_groups` on `groups`.`id` = `users_groups`.`group_id` where `users_groups`.`user_id` = '1'400μs
select * from `settings` where `settings`.`id` = '1' limit 1510μs
select * from `throttle` where `user_id` = '1' limit 1300μs
select * from `throttle` where `user_id` = '1' limit 1330μs
select * from `throttle` where `user_id` = '1' limit 1300μs
select * from `throttle` where `user_id` = '1' limit 1

**Comment by snipe** [comment]
Are you running this over SSL?

**Comment by DavidR-ZA** [comment]
No. This is an internal site not accessible to the outside world.

http://snipe.ks.local/admin/users

**Comment by snipe** [comment]
Look at the url in that debug console. It's trying to pull up https. You need to change your URL in `app/config/production/app.php` to use the http protocol, instead of https.

**Comment by DavidR-ZA** [comment]
OK I have edited that.
Trying again.

**Comment by DavidR-ZA** [comment]
No change on the people page but picked up this error on the /hardware/create page:
![image](https://cloud.githubusercontent.com/assets/17966170/13946080/f3eb5cb8-f01a-11e5-8d08-832d25340118.png)

**Comment by snipe** [comment]
Can you go to http://ks-vm-web.ks.local/api/statuslabel/1/deployable and tell me what you see? Also, can you confirm 100% that http://ks-vm-web.ks.local is your DNS name for this installation?

**Comment by snipe** [comment]
My guess here is that your `app/config/production/app.php` URL setting is still not correct. Note the Access-Control-Allow-Origin error above? If the domain you're on and the domain it's trying to pull the ajax call from are the same, you wouldn't see that error.

**Comment by DavidR-ZA** [comment]
You are good at this!
OK I had to change the server name because I added another server to the site so the snipe server had to be snipe.ks.local.
That is now correct in app.php
The add asset is now showing the list of users as it should.
Still no imported users in People page, though

**Comment by snipe** [comment]
Try hitting the refresh button in the table on the people view, or clicking on one of the table headers to re-sort, and tell me what happens, if anything.

**Comment by DavidR-ZA** [comment]
No - I have been refreshing the page regularly.

**Comment by snipe** [comment]
(Also make sure the search box on the People listing page is empty.)

**Comment by snipe** [comment]
Not the page, the table. There is a refresh icon in the table.

**Comment by DavidR-ZA** [comment]
No  - nothing in the search box.

**Comment by snipe** [comment]
Did you click on the table refresh (not the browser refresh)?

**Comment by DavidR-ZA** [comment]
Sorry - I meant refreshing the data with the refresh button.

**Comment by snipe** [comment]
Okay, when you click on the refresh table icon, I want you to click on the new dropdown in the bottom debug bar. When you first load the People page, it won't show you that bottom dropdown, but when you click refresh, it will make another ajax call, so you'll see a dropdown appear in the debug bar. 

<img width="1251" alt="screen shot 2016-03-22 at 2 05 12 am" src="https://cloud.githubusercontent.com/assets/197404/13946569/a6326950-efd2-11e5-8b6a-1c0222c2fd9c.png">

I want you to click on Queries in that bottom debug bar and tell me all of what you see there.

**Comment by snipe** [comment]
Also go to  http://snipe.ks.local/api/users/list and tell me what you see.

**Comment by DavidR-ZA** [comment]
This is what I get when I try to got to http://snipe.ks.local/api/users/list

{"total":2,"rows":[{"id":"1426","checkbox":"<div class=\"text-center hidden-xs hidden-sm\"><input type=\"checkbox\" name=\"edit_user[1426]\" class=\"one_required\"><\/div>","name":"<a title=\"Test Surname\" href=\"..\/admin\/users\/1426\/view\">Test Surname<\/a>","email":"<a href=\"mailto:test@kyalamischools.org\" class=\"hidden-md hidden-lg\">test@kyalamischools.org<\/a><a href=\"mailto:test@kyalamischools.org\" class=\"hidden-xs hidden-sm\"><i class=\"fa fa-envelope\"><\/i><\/a><\/span>","username":"Test","location":"SSC","manager":"<a title=\"David Rogers\" href=\"users\/1\/view\">David Rogers<\/a>","assets":0,"employee_num":"","licenses":0,"accessories":0,"consumables":0,"groups":"<a href=\"http:\/\/snipe.ks.local\/admin\/groups\/3\/edit\" class=\"label  label-default\">Users<\/a> ","notes":"","activated":"<i class=\"fa fa-check\"><\/i>","actions":"<nobr><a href=\"http:\/\/snipe.ks.local\/admin\/users\/1426\/edit\" class=\"btn btn-warning btn-sm\"><i class=\"fa fa-pencil icon-white\"><\/i><\/a> <a data-html=\"false\" class=\"btn delete-asset btn-danger btn-sm\" data-toggle=\"modal\" href=\"http:\/\/snipe.ks.local\/admin\/users\/1426\/delete\" data-content=\"Are you sure you wish to delete this user?\" data-title=\"Delete Test?\" onClick=\"return false;\"><i class=\"fa fa-trash icon-white\"><\/i><\/a> <\/nobr>","companyName":"Kyalami Schools"},{"id":"1","checkbox":"<div class=\"text-center hidden-xs hidden-sm\"><input type=\"checkbox\" name=\"edit_user[1]\" class=\"one_required\"><\/div>","name":"<a title=\"David Rogers\" href=\"..\/admin\/users\/1\/view\">David Rogers<\/a>","email":"<a href=\"mailto:david@kyalamischools.org\" class=\"hidden-md hidden-lg\">david@kyalamischools.org<\/a><a href=\"mailto:david@kyalamischools.org\" class=\"hidden-xs hidden-sm\"><i class=\"fa fa-envelope\"><\/i><\/a><\/span>","username":"david","location":"SSC","manager":"","assets":0,"employee_num":"","licenses":0,"accessories":0,"consumables":0,"groups":"<a href=\"http:\/\/snipe.ks.local\/admin\/groups\/1\/edit\" class=\"label  label-default\">Admin<\/a> ","notes":"Generated on install","activated":"<i class=\"fa fa-check\"><\/i>","actions":"<nobr><a href=\"http:\/\/snipe.ks.local\/admin\/users\/1\/edit\" class=\"btn btn-warning btn-sm\"><i class=\"fa fa-pencil icon-white\"><\/i><\/a>  <span class=\"btn delete-asset btn-danger btn-sm disabled\"><i class=\"fa fa-trash icon-white\"><\/i><\/span><\/nobr>","companyName":"Kyalami Schools"}]}

**Comment by snipe** [comment]
Do you have access to your database, like via phpmyadmin or something? If so, can you take a screenshot of the data in there? (Obscure email addresses if you can)

**Comment by DavidR-ZA** [comment]
These are the queries from the people page.

select * from `users` where `users`.`deleted_at` is null and `users`.`id` = '1' limit 1500μs
select * from `throttle` where `user_id` = '1' limit 1300μs
select `groups`.*, `users_groups`.`user_id` as `pivot_user_id`, `users_groups`.`group_id` as `pivot_group_id` from `groups` inner join `users_groups` on `groups`.`id` = `users_groups`.`group_id` where `users_groups`.`user_id` = '1'340μs
select * from `settings` where `settings`.`id` = '1' limit 1490μs
select count(*) as aggregate from `users` where `users`.`deleted_at` is null and `company_id` = '1' order by `first_name` asc290μs
select `users`.`id`, `users`.`employee_num`, `users`.`email`, `users`.`username`, `users`.`location_id`, `users`.`manager_id`, `users`.`first_name`, `users`.`last_name`, `users`.`created_at`, `users`.`notes`, `users`.`company_id`, `users`.`deleted_at`, `users`.`activated` from `users` where `users`.`deleted_at` is null and `company_id` = '1' order by `first_name` asc limit 50 offset 0350μs
select * from `assets` where `company_id` = '1' and `assets`.`assigned_to` in ('1', '1426')390μs
select `accessories`.*, `accessories_users`.`assigned_to` as `pivot_assigned_to`, `accessories_users`.`accessory_id` as `pivot_accessory_id`, `accessories_users`.`id` as `pivot_id` from `accessories` inner join `accessories_users` on `accessories`.`id` = `accessories_users`.`accessory_id` where `company_id` = '1' and `accessories_users`.`assigned_to` in ('1', '1426')380μs
select `consumables`.*, `consumables_users`.`assigned_to` as `pivot_assigned_to`, `consumables_users`.`consumable_id` as `pivot_consumable_id`, `consumables_users`.`id` as `pivot_id` from `consumables` inner join `consumables_users` on `consumables`.`id` = `consumables_users`.`consumable_id` where `company_id` = '1' and `consumables_users`.`assigned_to` in ('1', '1426')390μs
select `licenses`.*, `license_seats`.`assigned_to` as `pivot_assigned_to`, `license_seats`.`license_id` as `pivot_license_id`, `license_seats`.`id` as `pivot_id` from `licenses` inner join `license_seats` on `licenses`.`id` = `license_seats`.`license_id` where `licenses`.`deleted_at` is null and `company_id` = '1' and `license_seats`.`assigned_to` in ('1', '1426')430μs
select * from `users` where `users`.`id` in ('1')380μs
select * from `throttle` where `throttle`.`user_id` in ('1', '1426')300μs
select `groups`.*, `users_groups`.`user_id` as `pivot_user_id`, `users_groups`.`group_id` as `pivot_group_id` from `groups` inner join `users_groups` on `groups`.`id` = `users_groups`.`group_id` where `users_groups`.`user_id` in ('1', '1426')340μs
select * from `locations` where `locations`.`id` in ('1')330μs
select * from `companies` where `companies`.`id` in ('1')

**Comment by snipe** [comment]
Are you using full company support? (Check in Admin > Settings)

**Comment by DavidR-ZA** [comment]
![image](https://cloud.githubusercontent.com/assets/17966170/13946755/17ff56fa-f01f-11e5-86b1-44f5d0bd3874.png)

This is my user table from HeidiSQL

**Comment by DavidR-ZA** [comment]
I do have full Multiple companies Support set to Yes.

**Comment by snipe** [comment]
Do you actually need it turned on? I can't see from your screenshot, but I'm guessing your imported LDAP users don't have companies associated with them. If you don't need full company support, turn it off, as it will restrict what users and admins can see.

**Comment by DavidR-ZA** [comment]
OK Removed that and immediately all users appeared.

[Very happy smiley face emoticon]

**Comment by snipe** [comment]
Glad you're sorted.

**Comment by DavidR-ZA** [comment]
My location determines which company they are from.
Can I limit my technicians to one location only?

**Comment by snipe** [comment]
Not currently, but it's something that's planned.

**Comment by DavidR-ZA** [comment]
OK I will add company data to AD at each site.
Thanks again

**Comment by DavidR-ZA** [comment]
In the LDAP settings there is no option to set the mapping for the company field. I did have company settings in my AD data - it was not set to match the company names in Snipe but I edited my companies in Snipe to match what LDAP should bring in. However the import process does not import  the company field from AD.
I ran this query on the database
update users set company_id=location_id
because all my locations matched the companies and now every user has a company associated with them.
It would have been nicer though if the LDAP import supported all the fields used in Snipe.
